### ISSUE TITLE
Add trim utility

### ISSUE THREAD

**Issue by 3noch** [ndmitchell/shake#278]
Haskell's lack of a core `trim` function for `String` is ridiculous. I know it's easy to define one yourself, but folks from other ecosystems don't expect that such a useful, common function would be missing from standard headers. It's especially helpful when using the stdout of a command in further path manipulations. But often stdout out comes with leading and trailing whitespace.

**Comment by ndmitchell** [comment]
It's in the extra library: https://hackage.haskell.org/package/extra-1.3/docs/Data-List-Extra.html#v:trim

**Comment by 3noch** [comment]
You don't think it's worth re-exporting?

**Comment by ndmitchell** [comment]
The problem with re exporting is I'd need to decide what to reexport (it could end up being quite a lot) and it ties the packages closer together. Maybe a page listing some common useful functions such as trim and others? And giving some instructions on how to use the packages together?

**Comment by 3noch** [comment]
I suppose "findability" really is the problem at this point. Adding in `extra` is easy enough. It's just that searching for `trim` on Hackage is not going to limit your results to just the ones that are already brought in by Shake. Things like those in `extra` are annoying precisely because they are often too small to warrant a new dependency on their own. :(

**Comment by ndmitchell** [comment]
Indeed, hopefully putting them all together and directing people to them will be enough. Let's try that first.

**Comment by 3noch** [comment]
I'm fine with that. Now that I have the answer, I'm rolling. But perhaps I'm not alone.

**Comment by 3noch** [comment]
Thank you, by the way.

**Comment by ndmitchell** [comment]
I'll leave this ticket to add documentation somewhere - perhaps the "Haskell Zone" of the user manual.

**Comment by ndmitchell** [comment]
Also note #266 is asking similar questions.

**Comment by ndmitchell** [comment]
Added to the FAQ, under the encompassing title of "string manipulation" functions - which do seem to be the ones found in many languages and missing in Haskell.

**Comment by 3noch** [comment]
Great! While we're on the topic, perhaps you could also point us to a `leftpad` function. :P

**Comment by ndmitchell** [comment]
Hmm, actually the extra library misses a padding function. Maybe that's something that needs resolving...

**Comment by 3noch** [comment]
No need to clutter with the `extra` with that, it could just be its own package.

**Comment by ndmitchell** [comment]
If someone creates such a useful package, I guess I'd have no choice but to link to it.

**Comment by joneshf** [comment]
Could there be a `CmdOption` for `TrimStdout` or some data type that has a `CmdResult` instance? Forgetting to trim whitespace around `Stdout` leads to hard to track down bugs.

**Comment by ndmitchell** [comment]
@joneshf is it really that common a problem? I've rarely encountered it. And also do you trim each line, trim whitespace around lines etc? Is calling `trim` after that hard? Does having `TrimStdout` help, since if you forget to put in trim, you're likely to forget to pass `TrimStdout` too?

**Comment by joneshf** [comment]
I've not been hit by this often (because I don't usually need to process `Stdout`), but I think every time I've tried to process `Stdout`, I've been caught out by forgetting to trim a trailing newline.

Just last night, I spent two hours trying to figure out why `needed` wasn't working within an oracle. It was because I was calling `needed` on a file path like `"foo/bar/baz\n"` and not `"foo/bar/baz"`. I even remembered that I'd been caught out by newlines in the past. Printing the path with `putNormal` and friends didn't seem like there was anything wrong until also calling `show` on the path showed the trailing newline.

It's not that calling `trim` is hard, it's that there's nothing available within `shake` to do it. Since there's nothing in the package yet there's many helpers for other edge cases in the package, it's hard to even guess that trailing newlines would be a thing to deal with.

I think my issue is that I'm expecting unix-like behavior where dealing with trailing newlines from stdout isn't really a thing. It's fine for the behavior to be different, it's just hard when there's nothing within `shake` to deal with the different behavior.

At the end of the day, it's not that common of a thing for me to process `Stdout`, so it's not that big a pain point for me. I just wanted to raise my hand that even someone that's been using shake for a few months still gets caught out by this. It had been a couple years since the last response on the issue and sometimes thoughts change. If the answer is still to pull in another package to process `Stdout`, it's not terrible to just do that.
